routing/bicycle_directions.cpp

```diff
@@ -0,0 +1,216 @@
+    graph.GetOutgoingEdges(currJunction, outgoingEdges);
+    graph.GetIngoingEdges(currJunction, ingoingEdges);
+
+    AdjacentEdges adjacentEdges = AdjacentEdges(ingoingEdges.size());
+    // Outgoing edge angle is not used for bicyle routing.
+    adjacentEdges.m_outgoingTurns.isCandidatesAngleValid = false;
+    adjacentEdges.m_outgoingTurns.candidates.reserve(outgoingEdges.size());
+    ASSERT_EQUAL(routeEdges.size(), pathSize - 1, ());
+    FeatureID const inEdgeFeatureId = routeEdges[i - 1].GetFeatureId();
+
+    for (auto const & edge : outgoingEdges)
+    {
+      auto const & outFeatureId = edge.GetFeatureId();
+      // Checking for if |edge| is a fake edge.
+      if (!outFeatureId.IsValid())
+        continue;
+      adjacentEdges.m_outgoingTurns.candidates.emplace_back(0. /* angle */, outFeatureId.m_index,
+                                                            GetHighwayClass(outFeatureId));
+    }
+
+    LoadedPathSegment pathSegment;
+    if (inEdgeFeatureId.IsValid())
+      LoadPathGeometry(inEdgeFeatureId, {prevJunction.GetPoint(), currJunction.GetPoint()}, pathSegment);
+
+    m_adjacentEdges.insert(make_pair(inEdgeFeatureId.m_index, move(adjacentEdges)));
+    m_pathSegments.push_back(move(pathSegment));
+  }
+
+  RoutingResultGraph resultGraph(routeEdges, m_adjacentEdges, m_pathSegments);
+  RouterDelegate delegate;
+  Route::TTimes turnAnnotationTimes;
+  Route::TStreets streetNames;
+  MakeTurnAnnotation(resultGraph, delegate, routeGeometry, turns, turnAnnotationTimes, streetNames);
+}
+
+void BicycleDirectionsEngine::UpdateFeatureLoaderGuardIfNeeded(Index const & index, MwmSet::MwmId const & mwmId)
+{
+  if (!m_featuresLoaderGuard || mwmId != m_mwmIdFeaturesLoaderGuard)
+    m_featuresLoaderGuard.reset(new Index::FeaturesLoaderGuard(index, mwmId));
+}
+
+ftypes::HighwayClass BicycleDirectionsEngine::GetHighwayClass(FeatureID const & featureId)
+{
+  ftypes::HighwayClass highWayClass = ftypes::HighwayClass::Undefined;
+  MwmSet::MwmId const & mwmId = featureId.m_mwmId;
+  uint32_t const featureIndex = featureId.m_index;
+
+  FeatureType ft;
+  UpdateFeatureLoaderGuardIfNeeded(m_index, mwmId);
+  m_featuresLoaderGuard->GetFeatureByIndex(featureIndex, ft);
+  highWayClass = ftypes::GetHighwayClass(ft);
+  ASSERT_NOT_EQUAL(highWayClass, ftypes::HighwayClass::Error, ());
+  ASSERT_NOT_EQUAL(highWayClass, ftypes::HighwayClass::Undefined, ());
+  return highWayClass;
+}
+
+void BicycleDirectionsEngine::LoadPathGeometry(FeatureID const & featureId, vector<m2::PointD> const & path,
+                                               LoadedPathSegment & pathSegment)
+{
+  pathSegment.Clear();
+
+  MwmSet::MwmId const & mwmId = featureId.m_mwmId;
+  if (!featureId.IsValid())
+  {
+    ASSERT(false, ());
+    return;
+  }
+
+  FeatureType ft;
+  UpdateFeatureLoaderGuardIfNeeded(m_index, mwmId);
+  m_featuresLoaderGuard->GetFeatureByIndex(featureId.m_index, ft);
+  pathSegment.m_highwayClass =  ftypes::GetHighwayClass(ft);
```

Comment:
```c++
auto const cls = ftypes::GetHightwayClass(ft);
ASSERT_NOT_EQUAL(...);
...
pathSegment.m_highwayClass = cls;
```
I mean, it's better to initialize all fields of pathSegment at once. So, it won't be touched in case of errors, or completely initialized if everything is ok.